onnxruntime/core/optimizer/matmul_transpose_fusion.cc

```diff
@@ -86,6 +86,64 @@ static size_t UpdateConsumerCount(Graph& graph, NodeArg* target, std::unordered_
   }
 }
 
+static Node* GetTransposeNodeFromCast(Graph& graph, Node* cast) {
+  /* Interchange Cast and Transpose nodes in the graph and return Transpose node if possible
+  *  Requirements to interchange Cast and Transpose nodes changing the order of the operations.
+  *  1. Both Cast and Transpose are one-output nodes (assuming both have one-input only)
+  *  2. Transpose only feeds the Cast node (and no other node)
+  *  3. Cast only feeds the MalMul node (and no other node)
+  */
+
+  ORT_ENFORCE(cast != nullptr);
+  auto transpose = GetTransposeNodeFromOutput(graph, *cast->MutableInputDefs()[0]);
+  if (transpose == nullptr || cast->GetOutputEdgesCount() !=1 || transpose->GetOutputEdgesCount() !=1) {
+    return nullptr;
+  }
+  ORT_ENFORCE(cast->GetOutputEdgesCount() ==1 && transpose->GetOutputEdgesCount() <= 1);
+  NodeArg* cast_input = cast->MutableInputDefs()[0];
+  NodeArg* cast_output = cast->MutableOutputDefs()[0];
+  NodeArg* transpose_input = transpose->MutableInputDefs()[0];
+  NodeArg* transpose_output = transpose->MutableOutputDefs()[0];
+  ORT_ENFORCE(cast_input == transpose_output);
```

Comment:
since cast is the only consumer of transpose(output count check and transpose is obtained from cast), this check is redundant